File: rnn/matlab/rnn.m
function rnn = RNN(inputDim, hiddenDim, outputDim)

  % Initialize weights and biases
  weights = rand(inputDim, hiddenDim) - 0.5;
  biases = rand(hiddenDim) - 0.5;

  % Initialize hidden state
  h = zeros(hiddenDim, 1);

end

function rnn = train(rnn, data)

  % Train the RNN model
  for sequence in data

    for t in sequence

      % Compute the hidden state
      z = 0;
      for i in 1:inputDim
        z += data(t, i) * weights(i, h(end));
      end
      h = [h tanh(z + biases(end))];
    end

  end

end

function predictions = predict(rnn, data)

  % Make predictions on the given data
  predictions = [];
  h = zeros(hiddenDim, 1);
  for t in data

    % Compute the hidden state
    z = 0;
    for i in 1:inputDim
      z += data(t, i) * weights(i, h(end));
    end
    h = [h tanh(z + biases(end))];

    % Add the prediction to the list of predictions
    predictions = [predictions h(end)];
  end

  return predictions;

end

File: rnn/ruby/rnn.rb
class RNN

    def initialize(input_dim, hidden_dim, output_dim)
      @input_dim = input_dim
      @hidden_dim = hidden_dim
      @output_dim = output_dim
  
      @weights = Array.new(@input_dim) { Array.new(@hidden_dim) { rand - 0.5 } }
      @biases = Array.new(@hidden_dim) { rand - 0.5 }
  
      @h = Array.new(@hidden_dim)
    end
  
    def train(data)
      data.each do |sequence|
        sequence.each do |t|
          z = 0
          for i in 0...@input_dim
            z += t * @weights[i][@h.size]
          end
          @h.push(Math.tanh(z + @biases[@h.size]))
        end
      end
    end
  
    def predict(data)
      predictions = []
      @h = Array.new(@hidden_dim)
      data.each do |t|
        z = 0
        for i in 0...@input_dim
          z += t * @weights[i][@h.size]
        end
        @h.push(Math.tanh(z + @biases[@h.size]))
        predictions.push(@h.last)
      end
      return predictions
    end
  end
  